### PR TITLE
Migrate to zg from ziglyph

### PR DESCRIPTION
**Pull request by jacobsandlund**
This PR replaces `ziglyph` with `zg`, addressing https://github.com/ghostty-org/ghostty/issues/5694.

Note that this uses the [`jacob` branch from my fork of zg](https://codeberg.org/jacobsandlund/zg/src/branch/jacob) on codeberg, as there were a number of PRs I needed to make to `zg` to make it work for Ghostty: [#75](https://codeberg.org/atman/zg/pulls/75), [#76](https://codeberg.org/atman/zg/pulls/76), [#77](https://codeberg.org/atman/zg/pulls/77), [#78](https://codeberg.org/atman/zg/pulls/78), [#79](https://codeberg.org/atman/zg/pulls/79), [#1 on my fork](https://codeberg.org/jacobsandlund/zg/pulls/1) [#2 on my fork](https://codeberg.org/jacobsandlund/zg/pulls/2), and one more PR I'll make once some of these are merged.

~As discussed in the issue thread, this initializes the `zg` global constant state inside `global`. This does require initializing the `zg` portion of the global state to make tests pass, so a number of tests have a little bit more setup.~

**edit:** This PR now uses an even-more-modified version of `zg` that changes the API to basic functions using constant data internally.

This also adds a `unicode-test` executable, built with `zig build -Doptimize=ReleaseFast -Demit-unicode-test`, documented in `src/unicode/main.zig`. I used this to verify the implementation. In particular `./zig-out/bin/unicode-test class old` showed how 84 code points changed boundary classes with `zg` using Unicode 16.0.

### CODE REVIEW

File: src/input/Binding.zig [hunk 1]
```diff
@@ -1551,15 +1553,17 @@ pub const Trigger = struct {
     /// in more codepoints so we need to use a 3 element array.
     fn foldedCodepoint(cp: u21) [3]u21 {
         // ASCII fast path
-        if (ziglyph.letter.isAsciiLetter(cp)) {
-            return .{ ziglyph.letter.toLower(cp), 0, 0 };
+        if (('A' <= cp and cp <= 'Z') or ('a' <= cp and cp <= 'z')) {
```
Comment (jacobsandlund) [inline src/input/Binding.zig:1567]: zg lacks this `isAsciiLetter`, but it's a simple check so i didn't PR adding it

File: src/renderer/cell.zig [hunk 2]
```diff
@@ -246,8 +247,10 @@ pub fn fgMode(
             const cell = cell_pin.rowAndCell().cell;
             const cp = cell.codepoint();
 
-            if (!ziglyph.general_category.isPrivateUse(cp) and
-                !ziglyph.blocks.isDingbats(cp))
+            // If it's not Private Use (Co) or Dingbats (0x2700-0x27bf), use
+            // normal mode.
+            if (GeneralCategories.gc(zg.general_categories, cp) != .Co and
+                !(cp >= 0x2700 and cp <= 0x27bf))
```
Comment (jacobsandlund) [inline src/renderer/cell.zig]: `zg` doesn't have a module replacing `ziglyph.blocks`, but since we're only using it here, I didn't PR adding it

File: src/simd/codepoint_width.zig [hunk 3]
```diff
@@ -19,27 +21,41 @@ test "codepointWidth basic" {
...
-//             std.log.warn("mismatch cp=U+{x} simd={} zg={}", .{ cp, simd, zg });
-//             try testing.expect(false);
-//         }
-//     }
-// }
+//            if (cp == 0x890) {
+//                try testing.expectEqual(@as(i8, 0), simd);
+//                try testing.expectEqual(@as(i8, 1), zg_width);
+//                break :mismatch;
+//            }
```
Comment (jacobsandlund) [inline src/simd/codepoint_width.zig:47]: there's actually a bunch more failures when i started to try to run this, but in the `unicode-test` exe I verified this PR doesn't change our width calculation.

File: src/terminal/Terminal.zig [hunk 4]
```diff
@@ -344,7 +344,7 @@ pub fn print(self: *Terminal, c: u21) !void {
             // VS15 makes it narrow.
             if (c == 0xFE0F or c == 0xFE0E) {
                 // This only applies to emoji
-                const prev_props = unicode.getProperties(prev.cell.content.codepoint);
+                const prev_props = unicode.table.get(prev.cell.content.codepoint);
```
Comment (jacobsandlund) [inline src/terminal/Terminal.zig:347]: `getProperties` (aka `props.get`) now relies on `DisplayWidth`, which we don't pull in as a dependency for the main executable/library. I switched this to use our table data for the properties instead.

File: src/unicode/main.zig [hunk 5]
```diff
@@ -1,13 +1,238 @@
...
+///     class:    this verifies the table grapheme boundary classes match
+///     break:    this will verify the grapheme break implementation. This
+///               iterates over billions of codepoints so it is SLOW.
+///
+///     old:      compare against old implementation
+///     zg:       compare against zg
+///     ziglyph:  compare against ziglyph
+///
+/// Note: To disable/enable `old` comparisons, (un)comment sections of these
+/// files (search for "old"):
```
Comment (jacobsandlund) [inline src/unicode/main.zig:25]: I could comment out all the "old" comparison code before merging, if that's preferred. It was useful to verify this PR.

**Review by mitchellh — COMMENTED** [review]
Thank you, I still need to do a proper review. When I looked into this, the Zg initialization time added a significant amount of time to the startup. Did you measure this? I was trying to look for a way to defer initialization as long as possible -- or to get it done on a separate thread -- so initial startup wasn't blocked. But I don't recall the exact numbers or if they improved.

**Comment by jacobsandlund** [comment]
Good call, it does seem to increase the global init time from 0.6 ms to 7 ms on my machine. Let me think about deferring that or moving that to a thread and I'll update the PR.

Before:

```
[global init time] start=1751723204456874us duration=639us
```

With zg:

```
[global init time] start=1751722937788100us duration=7132us
```

**Comment by mitchellh** [comment]
Okay yeah, that's pretty significant. I think when I recorded the full startup time that made up a significant portion which is kind of nasty. I think because it has to do decompression. The challenge with putting it on a thread is the concurrency safety of it all.

**Comment by jacobsandlund** [comment]
I took a stab at rewriting zg's code gen from compressed binary to uncompressed constants in a zig file (matching Ghostty's approach) and that was pretty easy. I also noticed they were initializing the `Normalization` module in `CaseFolding` and by eliminating that ~5ms of the ~7ms of the global initialization time went away. When I checked the binary size of libghostty.a after these two changes it went down! (I didn't have a chance to do a fair comparison of just removing compression, but I suspect it's an insignificant amount of savings.)

By doing this the API can be simplified to not require globally initialized runtime data, so I'll PR that on zg separately to the code gen changes.

Given that I'll be making at least two more major PRs on zg and my branch is getting further away from zg's main, I'm not sure how comfortable you all are using my fork's changes until they hopefully eventually get incorporated into zg main.

I'll push up these changes hopefully tomorrow or Monday!

**Comment by mitchellh** [comment]
> By doing this the API can be simplified to not require globally initialized runtime data, so I'll PR that on zg separately to the code gen changes.

This would be hugely preferred. 

Let's keep track of the zg upstream issues in #5694. I'd like to give them time to address the PRs. Since you've been interacting with that community, do you think they're likely to be accepted? If not, we'll have to reassess. Feel free to hop in Discord if you want a more real-time chat.

File: src/build/SharedDeps.zig [hunk 6]
```diff
@@ -411,11 +411,18 @@ pub fn add(
...
-    if (b.lazyDependency("ziglyph", .{
+    if (b.lazyDependency("zg", .{
         .target = target,
         .optimize = optimize,
     })) |dep| {
-        step.root_module.addImport("ziglyph", dep.module("ziglyph"));
+        if (self.config.emit_bench or self.config.emit_unicode_test) {
+            step.root_module.addImport("Graphemes", dep.module("Graphemes"));
+            step.root_module.addImport("DisplayWidth", dep.module("DisplayWidth"));
+        }
```
Comment (jacobsandlund) [inline src/build/SharedDeps.zig:421]: I tried moving these addImports into GhosttyBench and GhosttyUnicodeTest, but it didn't work for GhosttyBench for some reason.

File: src/unicode/grapheme.zig [hunk 7]
```diff
@@ -2,6 +2,7 @@ const std = @import("std");
 const props = @import("props.zig");
 const GraphemeBoundaryClass = props.GraphemeBoundaryClass;
 const table = props.table;
+const oldTable = props.oldTable;
```
Comment (jacobsandlund) [inline src/unicode/grapheme.zig:5]: See below about this `oldTable` and `oldGraphemeBreak` stuff.

**Comment by jacobsandlund** [comment]
> This would be hugely preferred.

done! well, on my branch anyway.

I've updated this PR to use a heavily modified `zg` that now uses simple functions using internal constant data. No more global state that needs to be initialized, no more giant runtime memory arrays.

**Review by mitchellh — CHANGES_REQUESTED** [review]
Overall, these changes look excellent. Thank you for doing the work. I want to wait and see where the dust settles upstream with `zg` PRs to see what our path forward is there.

I'd also like to make sure we write/run some benchmarks on ASCII/Unicode throughput to make sure nothing regressed here. I can help with that later.

**Comment by alanmoyano** [comment]
@jacobsandlund nice work! In the `zg` repo the author responded in two of your PRs [here](https://codeberg.org/atman/zg/pulls/75#issuecomment-5758844) and [here](https://codeberg.org/atman/zg/pulls/79#issuecomment-5758826) asking to make an issue to track the changes. Sorry for pinging, I don't have a codeberg account.

**Comment by jacobsandlund** [comment]
@alanmoyano Thanks! I'll respond there shortly, hopefully ~today after work~ ~**edit:** tomorrow morning.~ **edit:** finally done (7/21... I got sick).

Also my experience working with`zg` inspired me to try and write a tiny unicode library 😅 . Not to say we shouldn't use `zg` after the PRs get in, but I got curious to see how it'd be starting from scratch. I'll hop into the Discord likely this weekend rather than sharing here or on the issue.

> I'd also like to make sure we write/run some benchmarks on ASCII/Unicode throughput to make sure nothing regressed here. I can help with that later.

@mitchellh sounds great!

File: src/renderer/cell.zig [hunk 8]
```diff
@@ -248,7 +248,8 @@ pub fn constraintWidth(cell_pin: terminal.Pin) u2 {
         // We consider powerline glyphs whitespace.
         if (isPowerline(prev_cp)) break :prev;
 
-        if (ziglyph.general_category.isPrivateUse(prev_cp)) {
+        // If it's Private Use (Co) use 1 as the width.
+        if (GeneralCategories.gc(cp) == .Co) {
```
Comment (Atomk) [inline src/renderer/cell.zig:252]: `prev_cp` changed to `cp`, is this intentional? (not familiar with the codebase at all, was just randomly ~~procrastinating~~ looking at some PRs)

Comment (jacobsandlund) [reply src/renderer/cell.zig:252]: probably not intentional, so thanks for pointing this out. there are some conflicts on this bit of code that i've noticed in a different branch of mine. this pr is a bit blocked on upstream zg changes, and there are some alternatives being considered.

**Comment by mitchellh** [comment]
Closing this for now since there's no path to this getting merged. I appreciate the work here, and we'll keep in touch via Discord on your uucode work.
